executor/memtable_reader.go

```diff
@@ -696,3 +701,299 @@ func (e *clusterLogRetriever) close() error {
+	}
+	if e.extractor.EndTime == 0 {
+		return nil, errors.New("denied to scan hot regions, please specified the end time, such as `update_time < '2020-01-01 00:00:00'`")
+	}
+
+	historyHotRegionsRequest := &HistoryHotRegionsRequest{
+		StartTime:  e.extractor.StartTime,
+		EndTime:    e.extractor.EndTime,
+		RegionIDs:  e.extractor.RegionIDs,
+		StoreIDs:   e.extractor.StoreIDs,
+		PeerIDs:    e.extractor.PeerIDs,
+		IsLearners: e.extractor.IsLearners,
+		IsLeaders:  e.extractor.IsLeaders,
+	}
+
+	return e.startRetrieving(ctx, sctx, pdServers, historyHotRegionsRequest)
+}
+
+func (e *hotRegionsHistoryRetriver) startRetrieving(
+	ctx context.Context,
+	sctx sessionctx.Context,
+	pdServers []infoschema.ServerInfo,
+	req *HistoryHotRegionsRequest,
+) ([]chan hotRegionsResult, error) {
+
+	var results []chan hotRegionsResult
+	for _, srv := range pdServers {
+		for typ := range e.extractor.HotRegionTypes {
+			req.HotRegionTypes = []string{typ}
+			jsonBody, err := json.Marshal(req)
+			if err != nil {
+				return nil, err
+			}
+			body := bytes.NewBuffer(jsonBody)
+			ch := make(chan hotRegionsResult)
+			results = append(results, ch)
+			go func(ch chan hotRegionsResult, address string, body *bytes.Buffer) {
+				util.WithRecovery(func() {
+					defer close(ch)
+					url := fmt.Sprintf("%s://%s%s", util.InternalHTTPSchema(), address, pdapi.HotHistory)
+					req, err := http.NewRequest(http.MethodGet, url, body)
+					if err != nil {
+						ch <- hotRegionsResult{err: errors.Trace(err)}
+						return
+					}
+					req.Header.Add("PD-Allow-follower-handle", "true")
+					resp, err := util.InternalHTTPClient().Do(req)
+					if err != nil {
+						ch <- hotRegionsResult{err: errors.Trace(err)}
+						return
+					}
+					defer func() {
+						terror.Log(resp.Body.Close())
+					}()
+					if resp.StatusCode != http.StatusOK {
+						ch <- hotRegionsResult{err: errors.Errorf("request %s failed: %s", url, resp.Status)}
+						return
+					}
+					var historyHotRegions HistoryHotRegions
+					if err = json.NewDecoder(resp.Body).Decode(&historyHotRegions); err != nil {
+						ch <- hotRegionsResult{err: errors.Trace(err)}
+						return
+					}
+					ch <- hotRegionsResult{addr: address, messages: &historyHotRegions}
+				}, nil)
+			}(ch, srv.StatusAddr, body)
+		}
+	}
+	return results, nil
+}
+
+func (e *hotRegionsHistoryRetriver) retrieve(ctx context.Context, sctx sessionctx.Context) ([][]types.Datum, error) {
+	if e.extractor.SkipRequest || e.isDrained {
+		return nil, nil
+	}
+
+	if !e.retrieving {
+		e.retrieving = true
+		results, err := e.initialize(ctx, sctx)
+		if err != nil {
+			e.isDrained = true
+			return nil, err
+		}
+		// Initialize the heap
+		e.heap = &hotRegionsResponseHeap{}
+		for _, ch := range results {
+			result := <-ch
+			if result.err != nil {
+				sctx.GetSessionVars().StmtCtx.AppendWarning(result.err)
+			}
+			if result.err != nil || len(result.messages.HistoryHotRegion) == 0 {
+				continue
```

Comment:
move `continue` to line 886 and check `len(result.messages.HistoryHotRegion) > 0`?

Reply:
done.